remote_attestation/rust/src/handshaker.rs

```diff
@@ -329,40 +329,45 @@ pub struct ServerHandshaker {
     /// Collection of previously sent and received messages.
     /// Signed transcript is sent in messages to prevent replay attacks.
     transcript: Transcript,
-    /// Additional info about the server, including configuration information and proof of
-    /// inclusion in a verifiable log.
-    additional_info: Vec<u8>,
 }
 
 impl ServerHandshaker {
     /// Creates [`ServerHandshaker`] with `ServerHandshakerState::ExpectingClientIdentity`
     /// state.
-    pub fn new(behavior: AttestationBehavior, additional_info: Vec<u8>) -> Self {
+    pub fn new(behavior: AttestationBehavior) -> Self {
         Self {
             behavior,
             state: ServerHandshakerState::ExpectingClientHello,
             transcript: Transcript::new(),
-            additional_info,
         }
     }
 
     /// Processes incoming `message` and returns a serialized remote attestation message.
     /// If [`None`] is returned, then no messages should be sent out to the client.
-    pub fn next_step(&mut self, message: &[u8]) -> anyhow::Result<Option<Vec<u8>>> {
-        self.next_step_util(message).map_err(|error| {
-            self.state = ServerHandshakerState::Aborted;
-            error
-        })
+    pub fn next_step(
+        &mut self,
+        message: &[u8],
+        additional_info: Vec<u8>,
```

Comment:
I may be missing something, but what determines what to pass here, and at exactly which step? AFAICT the additional info is only used at one particular step, so we still need to pass something else for all the other steps? Looking at the tests below it seems that we do keep passing the same value, but that is only actually used in one.

Also the idea of this method is that we just pipe in messages from the underlying transport, and it should move along the various steps in the state machine (and eventually produce the encryptor), so adding this argument here seems to complicate the API.

I'll let @conradgrobler and @ipetr0v weigh in too.

Another thing to consider is that, in practice, in most protocols that have some space for additional info, only a fixed amount of additional data can be passed in, usually 256 bits, usually the result of a hashing function (rather than the actual data). So maybe another approach worth considering in order to reduce the memory usage would be to make this a fixed size array, but keep it in the constructor. Would that work?

Reply:
Yes, only used some for some steps, passing it always is still not optimal.

> what determines what to pass here
afaik the config stays unchanged after the runtime is initiated. Hence it'd always be a clone of the same vec. 

Which leads me to the suggestion I raised in the chat, and which be the optimal approach for memory: We create a single instance of the additional_info vec in the loader, and only pass down references (with a static lifetime).

Reply:
I think it's safe to assume it does not change, but I'm not sure that implies it can be `'static` -- maybe it can be coerced to a static value at runtime by "leaking" it? but I'm not sure that's idiomatic.

Reply:
Also see my other comment below, maybe the fix you are looking for is just a one line change there?

Reply:
How about using an Arc<Vec<u8>> rather than a static reference? Cloning it should be cheap, and it would remove the need to pass the additional info into every step.

Reply:
Not entirely convinced we need.want reference counting if this will never be dropped. But either way, wouldn't Rc be better since it's a single thread? https://doc.rust-lang.org/std/rc/struct.Rc.html

Reply:
I agree with @jul-sh that reference counting (atomic or not) seems superfluous here.

I think there is probably a solution that involves adding explicit lifetimes to the various structs involved, but I'm not sure it's worth optimizing for that yet.

Also just to correct myself on another point: it looks like this extra data is indeed only used as input to a hash function (so it ends up being fixed size in the wire format):

https://github.com/project-oak/oak/blob/1cf966842d26d0c63df9deffc37074852b88576f/remote_attestation/rust/src/handshaker.rs#L713

Anyways, since we know that this gets used only once, maybe using `mem::take` or `Option::take` may be another approach?

Reply:
If the Rust compiler can infer that the AttestationServer instance will always outlive the handshaker (which hopefully it should be able to do in spite of the async stream), the additional info passed to the handshaker could just be a reference e.g. `&[u8]`.

The suggestion for using reference counting was based on the possibility that the compiler might not be able to accurately infer the lifetimes due to the async stream.

Regarding the wireformat: the additional info is also included separately in the server identity message (https://github.com/project-oak/oak/blob/main/remote_attestation/rust/src/handshaker.rs#L456-L464) and serialised (https://github.com/project-oak/oak/blob/main/remote_attestation/rust/src/message.rs#L269)

Reply:
The compiler isn't able to infer that, storing a ref in the handshaker struct requires a lifetime annotation. It is able to infer lifetimes when passing the ref to methods, which this PR implements. 

If we really want the additional_info to be maintained in the handshaker struct (as opposed being passed to methods) it's either ref counting, or the explicit lifetime approach I suggested. 

My opinion is that passing to methods is fine. If we feel strongly otherwise imo we should use the explicit lifetime approach.  

We definitely do not want the status quo of cloning.

Reply:
Passing to the `next_step` method seems out of question for me, it breaks encapsulation of a method that is supposed to deal with a stream of opaque messages from the outside world. In fact eventually it may even implement `Sink<Vec<u8>>` or similar (at least conceptually).

What would this look like with the explicit lifetime annotation @jul-sh ?

Reply:
I don't really see it breaking encapsulation. It's just a question of whether this additional data (quite literally the variable name 😅) is passed on each call or initially and cloned as state. Happy to talk about it more though to understand your PoV. 

Anyways, with explicit lifetime annotations we would continue passing the additional info when the `Handshaker` is instantiated by `AttestationServer`. We use them to declare that `AttestationServer` always outlives `Handshaker`. That is the case (and always will be) but because of the async stream the compiler is unable to infer that implicitly.

Reply:
> Also the idea of this method is that we just pipe in messages from the underlying transport

I agree, I think having more arguments in the `next_step` function may be inconvenient.

Also, could we just have a shared reference to the `additional_info` instead? If the only problem is it being copied multiple times, shared reference could be fine, I think

Reply:
If passing it to the method is not idiomatic, we can make it a shared reference (std::sync::Arc) indeed. That would probably be easier than declaring the lifetimes. 

The only caveat is that as of #2556 the `oak_remote_attestation` crate is no_std compatible. This means we cannot use `std::sync::Arc` without breaking that. 

@conradgrobler Do we plan to support shared references in no_std in the future?

Reply:
You can use `alloc::sync::Arc`. We will always have an allocator, even in no_std environements.

Reply:
Thanks for the pointer, will use that. Also realized that for no_std we'll probs be moving this gRPC communication logic out of the trusted runtime anyways.

Reply:
Will we? I thought all this stuff will remain in the trusted binary, since it drives the remote attestation protocol.

Anyways, I still don't think we should change this method to accept the additional info param. Especially in the non-std, blocking model, this method should look like a `call` or `invoke` function from bytes to bytes (with some extra wrappers), so that we can compose it more nicely with the rest of the building blocks of the runtime. And at some point it should itself also accept another callback that it invokes with the decrypted data from remote attestation, which itself would follow the same paradigm.